src/dmd/statementsem.d

```diff
@@ -1670,306 +1761,250 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
-                    {
-                        fs.error("`opApply()` function for `%s` must return an `int`", tab.toChars());
-                        goto case Terror;
-                    }
-                }
-                else
-                {
-version (none)
-{
-                    if (global.params.vsafe)
-                    {
-                        message(loc, "To enforce `@safe`, the compiler allocates a closure unless `opApply()` uses `scope`");
-                    }
-                    flde.fd.tookAddressOf = 1;
-}
-else
-{
-                    if (global.params.vsafe)
-                        ++flde.fd.tookAddressOf;  // allocate a closure unless the opApply() uses 'scope'
-}
-                    assert(tab.ty == Tstruct || tab.ty == Tclass);
-                    assert(sapply);
-                    /* Call:
-                     *  aggr.apply(flde)
-                     */
-                    ec = new DotIdExp(loc, fs.aggr, sapply.ident);
-                    ec = new CallExp(loc, ec, flde);
-                    ec = ec.expressionSemantic(sc2);
-                    if (ec.op == TOK.error)
-                        goto case Terror;
-                    if (ec.type != Type.tint32)
-                    {
-                        fs.error("`opApply()` function for `%s` must return an `int`", tab.toChars());
-                        goto case Terror;
-                    }
-                }
-                e = Expression.combine(e, ec);
+    private static extern(D) Expression applyDelegate(ForeachStatement fs, Expression flde,
+                                                      Scope* sc2, Type tab)
+    {
+        Expression ec;
+        /* Call:
+         *      aggr(flde)
+         */
+        if (fs.aggr.op == TOK.delegate_ && (cast(DelegateExp)fs.aggr).func.isNested() &&
+            !(cast(DelegateExp)fs.aggr).func.needThis())
+        {
+            // https://issues.dlang.org/show_bug.cgi?id=3560
+            fs.aggr = (cast(DelegateExp)fs.aggr).e1;
+        }
+        ec = new CallExp(fs.loc, fs.aggr, flde);
+        ec = ec.expressionSemantic(sc2);
+        if (ec.op == TOK.error)
+            return null;
+        if (ec.type != Type.tint32)
+        {
+            fs.error("`opApply()` function for `%s` must return an `int`", tab.toChars());
+            return null;
+        }
+        return ec;
+    }
 
-                if (!fs.cases.dim)
-                {
-                    // Easy case, a clean exit from the loop
-                    e = new CastExp(loc, e, Type.tvoid); // https://issues.dlang.org/show_bug.cgi?id=13899
-                    s = new ExpStatement(loc, e);
-                }
-                else
-                {
-                    // Construct a switch statement around the return value
-                    // of the apply function.
-                    auto a = new Statements();
+    private static extern(D) Expression applyArray(ForeachStatement fs, Expression flde,
```

Comment:
Creating new functions with no Ddoc is not acceptable.

Reply:
ditto